backend/substrapp/tasks/task.py

```diff
@@ -0,0 +1,101 @@
+"""
+This file contains the main logic for executing a compute task:
+
+- Create execution context
+- Populate asset buffer
+- Loads assets from the asset buffer
+- **Execute the compute task**
+- Save the models/results
+- Teardown the context
+
+We also handle the retry logic here.
+"""
+import enum
+import pickle  # nosec B403
+from typing import Any
+
+import structlog
+from billiard.einfo import ExceptionInfo
+from celery import Task
+from django.conf import settings
+
+import orchestrator
+from substrapp.compute_tasks.compute_pod import delete_compute_plan_pods
+from substrapp.models import FailedAssetKind
+from substrapp.task_routing import WORKER_QUEUE
+from substrapp.tasks.tasks_asset_failure_report import store_asset_failure_report
+
+logger = structlog.get_logger(__name__)
+
+
+class FailableTask(Task):
+    asset_type: FailedAssetKind
+
+    # Celery does not provide unpacked arguments, we are doing it in `get_task_info`
+    def on_failure(
+        self, exc: Exception, task_id: str, args: tuple, kwargs: dict[str, Any], einfo: ExceptionInfo
+    ) -> None:
+        asset_key, channel_name = self.get_task_info(args, kwargs)
+        exception_pickled = pickle.dumps(exc)
+        store_asset_failure_report.apply_async(
+            args,
+            {
+                "asset_key": asset_key,
+                "asset_type": self.asset_type,
+                "channel_name": channel_name,
+                "exception_pickled": exception_pickled,
+            },
+            queue=WORKER_QUEUE,
+        )
+
+    def get_task_info(self, args: tuple, kwargs: dict) -> tuple[str, str]:
+        raise NotImplementedError()
+
+
+class ComputeTaskSteps(enum.Enum):
+    BUILD_IMAGE = "build_image"
+    PREPARE_INPUTS = "prepare_inputs"
+    TASK_EXECUTION = "task_execution"
+    SAVE_OUTPUTS = "save_outputs"
+
+
+class ComputeTask(FailableTask):
+    autoretry_for = settings.CELERY_TASK_AUTORETRY_FOR
+    max_retries = settings.CELERY_TASK_MAX_RETRIES
+    retry_backoff = settings.CELERY_TASK_RETRY_BACKOFF
+    retry_backoff_max = settings.CELERY_TASK_RETRY_BACKOFF_MAX
+    retry_jitter = settings.CELERY_TASK_RETRY_JITTER
+
+    asset_type = FailedAssetKind.FAILED_ASSET_COMPUTE_TASK
+
+    @property
+    def attempt(self) -> int:
+        return self.request.retries + 1  # type: ignore
+
+    # Celery does not provide unpacked arguments
+    def on_success(self, retval: dict[str, Any], task_id: str, args: tuple, kwargs: dict[str, Any]) -> None:
+        from django.db import close_old_connections
+
+        close_old_connections()
+
+    # Celery does not provide unpacked arguments, we are doing it in `get_task_info`
```

Comment:
I don't understand the comments about unpacking celery arguments and the relation to `get_task_info`, isn't what `split_args` is doing?